### PR TITLE
Comment with hopefully both correct and clearer setting of hsig

### PR DESCRIPTION
**Pull request by drallensmith**
This is to, I hope, help with #33. (Escaping the quote in the comment was needed in order for (x)emacs to properly do syntax coloring.) Admittedly, if the below is correct, then I don't see why one would then potentially add 0 (for hsig == 0) for self.pc[i] (why not simply decide whether or not to add `ccn * y[i]` outside the loop?), nor the use in `c1a = par.c1 * (1 - (1-hsig**2) * par.cc * (2-par.cc))`.

### CODE REVIEW

File: cma/purecma.py [hunk 1]
```diff
@@ -280,7 +280,7 @@ class CMAES(OOOptimizer):  # could also inherit from object
         print('best solution =', es.result[0])
 
         print('potentially better solution xmean =', es.result[5])
-        print("let's check f(xmean) = ", pcma.ff.elli(es.result[5]))
+        print("let\'s check f(xmean) = ", pcma.ff.elli(es.result[5]))
```
Comment (nikohansen) [inline cma/purecma.py]: what's wrong with the string `"'"`?

**Comment by nikohansen** [comment]
> Escaping the quote in the comment was needed in order for (x)emacs to properly do syntax coloring.

shouldn't this be considered as a bug in emacs?

**Comment by nikohansen** [comment]
Computation of `c1a` is meant to correct for the variance loss in `C` when `hsig` is zero. This should rarely matter, but IIRC I have seen one case where this made the difference between successful vs premature convergence. 

As long as the step-size is too small, any change of the covariance matrix is arguably unjustified because it comes from a "wrong" environment. Using repeatedly the "frozen" `pc` to update `C` may not at all be a good idea. It is probably comparatively easy to verify this empirically.

**Comment by drallensmith** [comment]
> shouldn't this be considered as a bug in emacs?

I can see both sides. It's actually provably impossible, IIRC, to have infinite levels of comments parsable by a DFA or NFA regular expression; xemacs is a general programming editor so doesn't have the Python specializations of, say, IDLE (which I've never gotten to work on my OS X machine).

> Computation of c1a is meant to correct for the variance loss in C when hsig is zero. This should rarely matter, but IIRC I have seen one case where this made the difference between successful vs premature convergence.

I can understand that part; I'm just confused by why a value that's always 0 or 1 is being put through `(1-hsig**2)`. Was it originally a more continuous value?

> As long as the step-size is too small,

As in if sigma is increasing quickly (`turn off rank-one accumulation when sigma increases quickly`)?

> Using repeatedly the "frozen" pc to update C may not at all be a good idea. It is probably comparatively easy to verify this empirically.

Any suggestions on good domains to test this in? If I have time to do that, I'd be happy to contribute any resulting code to be used as examples.

**Comment by drallensmith** [comment]
BTW, regarding using the "frozen" pc - actually, pc will not be "frozen" even if hsig is 0, assuming that par.cc is != 0. And are you meaning that it's uncertain whether it's a good idea to do: `self.C.addouter(self.pc, par.c1)  # C += c1 * pc * pc^T, so-called rank-one update` when hsig == 0?

**Comment by nikohansen** [comment]
Yes, step-size too small and `sigma` increases quickly are most of the time the same scenario, but I believe I chose my wording on purpose as I did. The sphere `cma.ff.sphere` with a far too small initial step-size is the standard way to test this.

> infinite levels of comments 

I only see one level and two methods, namely using single or double quotes.

> I'm just confused by why a value that's always 0 or 1 is being put through (1-hsig**2).

It indicates that `hsig` is considered as a standard deviation. Indeed it also suggests that the formula remains correct if `hsig` were to assume other values than it actually does.

> pc will not be "frozen" even if hsig is 0

that is why frozen is in quotes, its direction is frozen.

> And are you meaning that it's uncertain whether it's a good idea to do: self.C.addouter(self.pc, par.c1) # C += c1 * pc * pc^T, so-called rank-one update when hsig == 0?

I am pretty convinced that it is a bad idea in the unusual situation when `sigma` is (far) too small. I have probably spent a few hours to verify this, but I could always be mistaken.

**Comment by nikohansen** [comment]
On the code situation: I am here trying to spent as few lines as possible on parts of the code which are IMHO conceptually not very interesting or relevant.

**Comment by drallensmith** [comment]
Re sigma increases: I can see the possibility of objective functions that fluctuate very quickly in what overall stepsize is needed.

Re levels of comments: The docstring is itself in a triple quote (which additionally is unusual in Python in that it is allowed to stretch over multiple lines without `\` continuations).

In regards to the last, I trust (tongue in cheek) you will forgive me for finding portions of your code more interesting than you do?

File: cma/purecma.py [hunk 2]
```diff
@@ -179,19 +181,18 @@ def fmin(objective_fct, xstart, sigma,
...
 class CMAESParameters(object):
-    """static "internal" parameter setting for `CMAES`
-
-    """
+    """Static "internal" parameter setting for `CMAES`"""
     default_popsize = '4 + int(3 * log(N))'
     def __init__(self, N, popsize=None,
                  RecombinationWeights=None):
-        """set static, fixed "strategy" parameters once and for all.
+        """
```
Comment (nikohansen) [inline cma/purecma.py]: Is there a style guide which recommends a leading new line?

File: cma/purecma.py [hunk 3]
```diff
@@ -674,7 +674,7 @@ def legend_(*args, **kwargs):
         title_('Coordinate-wise STDs w/o sigma')
         grid(True)
         xlabel('iterations' + strpopsize)
-        _stdout.flush()
```
Comment (nikohansen) [inline cma/purecma.py]: it wasn't a typo, was it?

File: cma/purecma.py [hunk 4]
```diff
@@ -60,6 +60,7 @@
     from .recombination_weights import RecombinationWeights
 except (ImportError, ValueError):
     RecombinationWeights = None
+# XXX What is the below line doing?
```
Comment (nikohansen) [inline cma/purecma.py]: deleting namespace clutter

File: cma/purecma.py (hunk 3 above)
Comment (drallensmith) [reply cma/purecma.py]: Heh. No. It's that _stdout is supposed to be the original one, not whatever stdout has been renamed to (print will go to the latter).

File: cma/purecma.py (hunk 2 above)
Comment (drallensmith) [reply cma/purecma.py]: I had certainly thought it did, but apparently not; either with or without is considered fine, although the core libraries tend to use with a leading new line IIRC.

File: cma/purecma.py (hunk 4 above)
Comment (drallensmith) [reply cma/purecma.py]: OK, will note that.

File: cma/purecma.py (hunk 2 above)
Comment (drallensmith) [reply cma/purecma.py]: I'll go over the docstrings with reference to https://www.python.org/dev/peps/pep-0257/.

Comment (drallensmith) [reply cma/purecma.py]: One disadvantage of not using a leading new line is that one is supposed to fit a summary line as the first one (something I admittedly tend to forget), and it helps to make sure the line isn't so long that github won't display all of it.

**Comment by drallensmith** [comment]
On OS X, the doctest is actually failing currently:

python2 purecma.py
launching doctest...
**********************************************************************
File "purecma.py", line 1303, in __main__.test
Failed example:
    print("%s, %d" % (pcma.ff.rosenbrock(es.result[0]) < 1e13,
                      es.result[2]))
Expected:
    True, 1584
Got:
    True, 1560
**********************************************************************
1 items had failures:
   1 of  21 in __main__.test
***Test Failed*** 1 failures.
TestResults(failed=1, attempted=31)

File: cma/purecma.py [hunk 5]
```diff
@@ -681,12 +692,12 @@ def legend_(*args, **kwargs):
         CMAESDataLogger.plotted += 1
 
     def save(self, name=None):
-        """Save data to file `name` or ``self.filename``"""
+        """Save data to file `name` or ``self.filename``."""
```
Comment (nikohansen) [inline cma/purecma.py]: I use the missing dot as a convention that this is all there is in the docstring.

**Pull request closed without merging**